pipes.py

```diff
@@ -58,17 +61,18 @@ def is_colliding(self, game_state: FlappyNoleGameState):
         y_offset: int = round(character_pos[1] * -1)
         return CHARACTER_HITBOX.overlap(self.hitbox(game_state.screen_height), (x_offset, y_offset)) != None
 
+    # need to take into account the height of the image
     def tophalf(self, tick: int):
-        return pygame.Rect((self.visible_left_bound(tick), 0), (self.visible_width(tick), self.gap_start_pos))
+        return pygame.Rect((self.visible_left_bound(tick), 0 - PIPE_HEIGHT + self.gap_start_pos ), (self.visible_width(tick), self.gap_start_pos))
```

Comment:
Consider replacing the constant PIPE_HEIGHT with a simple call to `image.get_height()`. Declaring a variable  for height is redundant since we only ever use the height value once, right here.

Also hard-coding the value is redundant since we can just read it from the image object.

Reply:
I'd like to move this offset logic outside of the Pipe class and into the draw_pipes method. I think it is a better fit there. After all the only point of the offset is to crop the image, not actually calculate the position/placement of the pipes in the game world. In general I'd like to keep the pipe class decoupled from the rendering of the pipe itself. Separation of concerns, etc.

See the example I included below, which uses the Rect.move method.

<img width="1109" alt="Screen Shot 2021-11-05 at 12 22 02 AM" src="https://user-images.githubusercontent.com/9706506/140457925-eb36c951-a388-4307-8073-747d536c40ab.png">